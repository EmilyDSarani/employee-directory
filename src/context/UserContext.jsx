import { useContext, createContext, useState } from 'react'
import { getUser } from '../services/user';

//We will use this to set the user in context and have it be across the entire app.
//it is a big bucket of state that is shared across the entire app

//Create Context
const UserContext = createContext();

//Create Provider
const UserProvider = ({ children }) => {
    const currentUser = getUser()
    //when I load any page, its going to try to get the user from local storage
    //if that user exists, then set the email to user email
    const [user, setUser] = useState(currentUser?{email:currentUser.email} : {})

    //userAuth
    const login = (email, password) =>{
        const loginSuccessful =
        email === '' &&
        password === '';
        if(loginSuccessful) setUser(email);
        return loginSuccessful
    }
    //set logout
    const logout =(callback) =>{
        setUser(null);
        callback();
    };
    return <UserContext.Provider value={{user, setUser, login, logout}}>{children}</UserContext.Provider>
}
    //useUser Custom Hook
    const useUser = () => {
        const context = useContext(UserContext)

        if(context === undefined){
            throw new Error('useUser needs to be defined in UserContext Provider')
        }
        return context
    }


export {useUser, UserProvider}