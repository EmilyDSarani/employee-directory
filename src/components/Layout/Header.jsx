import { useUser } from "../../context/UserContext";
import { Link, useHistory } from "react-router-dom"
import { signOutUser } from "../../services/user";
import HeaderButton from "./HeaderButton";

//on click returns a redirect

export default function Header() {
    const { user, setUser } = useUser();
    const history = useHistory();

    // const handleLogOut = () => {
    //     signOutUser();
    //     setUser(null)
    //     history.push('/login');
    // };


    return (
        <>
         <header>
        <h2>
            ACME Inc. Employee Directory
        </h2>
        <p>
        {user?.email 
        ? `Logged in as ${user.email}` 
        : 'Not logged in'}
        <HeaderButton />
        </p>
         </header> 
        </>
    )
}
