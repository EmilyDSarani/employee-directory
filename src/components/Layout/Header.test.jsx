import {render, screen} from '@testing-library/react'
import { UserProvider } from '../../context/UserContext'
import { MemoryRouter } from 'react-router-dom'
import Header from './Header'

//PLEASE NOTE-- for mocking, you need to create a __mocks__ folder that will point the jest.mock stuff that is below to those folders. 
// then, you need to replace the content in those folders to match whatever you are mocking and put mockUser in front of some stuff in the UserContext for the hook.

jest.mock('../../context/UserContext.jsx')
jest.mock('../../services/user.js')

it('should display header when signed out', () => {
    const { container } = render(
        <UserProvider>
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        </UserProvider>
    );
    expect(container).toMatchSnapshot();
});
it('should display user email when signed in', () =>{
    const { container } = render(
        <UserProvider mockUser={{ id:1, email: 'example@jest.com'}}>
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        </UserProvider>
    );
screen.getByText('example@jest.com', { exact: false });
expect(container).toMatchSnapshot();
});