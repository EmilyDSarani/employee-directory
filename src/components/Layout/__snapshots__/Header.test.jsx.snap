// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display header when signed out 1`] = `
<div>
  <header>
    <h2>
      ACME Inc. Employee Directory
    </h2>
    <p>
      Not logged in
      <div>
        <a
          href="/login"
        >
          <button>
            Sign in
          </button>
        </a>
      </div>
    </p>
  </header>
</div>
`;

exports[`should display user email when signed in 1`] = `
<div>
  <header>
    <h2>
      ACME Inc. Employee Directory
    </h2>
    <p>
      Logged in as example@jest.com
      <div>
        <button>
          LogOut
        </button>
      </div>
    </p>
  </header>
</div>
`;
